### ISSUE TITLE
Snap segfaults on launch after upgrade

### ISSUE THREAD

**Issue by dpniel** [ubuntu/snapcraft-desktop-helpers#44]
When trying to launch dekko after an upgrade it is segfaulting on launch. I have tracked this down to the environment setup logic after an upgrade in the desktop-launch script, but I'm not really sure how to dig into this any further. It is easy to reproduce.

## Steps to reproduce

1. snap install --revision 22 ubuntu-app-platform ( in case Timo releases 33 to stable before you get to this)
2. snap install --edge --devmode --revision 65 dekko
3. snap connect dekko:platform ubuntu-app-platform:platform
4. run dekko and then close.
5. snap refresh --edge ubuntu-app-platform
6. snap refresh --edge --devmode dekko
7. snap connect dekko:platform ubuntu-app-platform:platform
8. Run dekko and it will segfault on launch
9. Delete `~/snap/dekko/119/.last_revision` so the desktop-launch doesn't follow the `needs_update` path
10. Run Dekko again and now it will launch

Let me know if there is anything else I can provide.

**Comment by sergiusens** [comment]
this procedure worked for me

**Comment by didrocks** [comment]
Same here, no crash on my end.

**Comment by sergiusens** [comment]
Let be clearer `rm ~/snap/dekko/119/.last_revision` made it work for me.

**Comment by didrocks** [comment]
Ok, I think the fact that I needed to disable/reenable it at step 7 to reconnect did screw up some parts of the testing procedure for me.
You didn't get the ackward "you need to reconnect" message after refreshing both snaps?

**Comment by didrocks** [comment]
Yeah, I'm still having after the second connect on step 7:
```
didrocks@tidus:~$ snap refresh --edge ubuntu-app-platform
ubuntu-app-platform (edge) 1 from 'canonical' refreshed
didrocks@tidus:~$ snap refresh --edge --devmode dekko
2017-02-08T09:39:03+01:00 INFO snap "dekko" has bad plugs or slots: unity8 (unknown interface)
dekko (edge) 0.9.7 from 'dekkoproject' refreshed
didrocks@tidus:~$ snap connect dekko:platform ubuntu-app-platform:platform
didrocks@tidus:~$ dekko 
You need to connect the ubuntu-app-platform package with your application 
to reuse shared assets, please run:
snap install ubuntu-app-platform
snap connect dekko:platform ubuntu-app-platform:platform
```

and snap interfaces show them as being connected. 
```
$ snap --version
snap    2.21
snapd   2.21
series  16
ubuntu  16.04
```
(latest in xenial-updates)

So, even if I can't reproduce it, I think I know what's happening. Just a hint with snap info:
snap info gives me for dekko:
```
  beta:    0.9.7 (82)  28MB devmode
  edge:    0.9.7 (119) 9MB  -
```

which is slightly a lie: upstream version is maybe the same, but the package version is clearly not, and I think as you are connecting to the new app-platform, you need the update script to run.

The comparison on "needs updating" is only made on what's available inside the snap yaml, which is the version. Here, the versions are the same (0.9.7) => so no update needed. You fix it by removing the version stamp, which then forces an update.
There isn't any access to the revision from within the snap if I'm correct. I think the only way, as you have a lot of your snap content structure changing due to this upgrade, to ensure version is different (like appending -1)

Making sense?

**Comment by sergiusens** [comment]
IIRC `SNAP_REVISION` is there

**Comment by didrocks** [comment]
Good point, I could use that to ensure that even if the same version is kept, any new revision will trigger an update.
Note that there is a cost on launch time, but agreed, that may be safer.

**Comment by didrocks** [comment]
Argh, checking again todayt, no, I was wrong, I'm using SNAP_REVISION already. So, it's not this. I wonder how the revision could matches though after the update, as != revisions is like removing the file.

Would be great for me to reproduce and not having this autoconnect issue first, do any of you got it? I've read multiple times this was a snapd bug and I'm surprised you don't get the same issue. Which version of snapd on what distro version are you using?

**Comment by didrocks** [comment]
I guess I just fixed it in 529e3290508b0e140f528339ec1db6906e9fc1d5. Please, reopen if that's not the case.
